File: testSuite/test-treeFilterLabels.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::IO::HDF5;
use PDL::NiceSlice;

# Run a Galacticus model to test node labeling functionality.
# Andrew Benson (07-July-2023)

# Ensure output directory exists.
system("mkdir -p outputs");

# Run model without checkpointing.
system("cd ..; ./Galacticus.exe testSuite/parameters/treeFilterLabels.xml"  );
die("FAILED: failed to run model")
    unless ( $? == 0 );

# Read model data.
my $data;
my $model = new PDL::IO::HDF5("outputs/treeFilterLabels.hdf5");
my $nodes = $model->group("Outputs/Output1/nodeData");
die("FAILED: label not output")
    unless ( grep {$_ eq "nodeLabelLMC"} $nodes->datasets() );
$data->{$_} = $nodes->dataset($_)->get()
	foreach ( "nodeIsIsolated", "nodeLabelLMC", "darkMatterProfileDMOVelocityMaximum", "basicTimeLastIsolated" );
my $hosts = which($data->{'nodeIsIsolated'} == 1);
my $lmcs  = which($data->{'nodeLabelLMC'  } == 1);
print "Found ".nelem($lmcs)." LMCs in ".nelem($hosts)." trees\n";
die("FAILED: LMCs not found in all trees")
    unless ( nelem($lmcs) >= nelem($hosts) );
die("FAILED: LMCs labelled prior to infall time")
    if ( any($data->{'basicTimeLastIsolated'              }->($lmcs) < 11.8) );
die("FAILED: LMCs labelled at low Vₘₐₓ")
    if ( any($data->{'darkMatterProfileDMOVelocityMaximum'}->($lmcs) < 55.0) );
print "SUCCESS: filtered tree labeling\n";

exit;

File: testSuite/test-treeFilterLabels.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run a Galacticus model to test node labeling functionality.
# Andrew Benson (ported to Python)

# Ensure output directory exists.
subprocess.run("mkdir -p outputs", shell=True)

# Run model.
status = subprocess.run("cd ..; ./Galacticus.exe testSuite/parameters/treeFilterLabels.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run model")
    sys.exit(0)

# Read model data.
with h5py.File("outputs/treeFilterLabels.hdf5", "r") as model:
    nodes = model["Outputs/Output1/nodeData"]
    if "nodeLabelLMC" not in nodes:
        print("FAILED: label not output")
        sys.exit(0)
    nodeIsIsolated      = nodes["nodeIsIsolated"][:]
    nodeLabelLMC        = nodes["nodeLabelLMC"][:]
    darkMatterVmax      = nodes["darkMatterProfileDMOVelocityMaximum"][:]
    basicTimeLastIsolated = nodes["basicTimeLastIsolated"][:]

hosts = np.where(nodeIsIsolated == 1)[0]
lmcs  = np.where(nodeLabelLMC  == 1)[0]
print(f"Found {len(lmcs)} LMCs in {len(hosts)} trees")

if len(lmcs) < len(hosts):
    print("FAILED: LMCs not found in all trees")
    sys.exit(0)

if np.any(basicTimeLastIsolated[lmcs] < 11.8):
    print("FAILED: LMCs labelled prior to infall time")
    sys.exit(0)

# Allow some tolerance in Vmax test as peak velocities are determined numerically for tidally-heated subhalos.
if np.any(darkMatterVmax[lmcs] < 54.0):
    print("FAILED: LMCs labelled at low Vmax")
    sys.exit(0)

print("SUCCESS: filtered tree labeling")
